docs/design-documents/20220309-opencdc.md

```diff
@@ -0,0 +1,350 @@
+- Replace the current OpenCDC record entirely by sending the parsed record further down the pipeline and discarding the original record.
+
+Note that this transform could only operate on the key or payload data at a time. If both the key and payload contain an OpenCDC record, then two chained transforms would need to be used to parse both.
+
+Note also that we could provide such transforms not only for OpenCDC but also for other formats (e.g. Debezium records).
+
+## Implementation options
+
+The format that we choose and define here wouldn't be constrained by a specific representation like JSON or protobuf. We would need a protobuf definition of the record to be able to move it from the plugin to Conduit through gRPC, but the record could just as well be formatted as JSON or another format.
+
+### Option 1
+
+We consider our needs, define OpenCDC from scratch and change our record to match OpenCDC.
+
+Note that since this is the recommended option we go into more details about how the record should change if we go ahead with this option.
+
+#### Pros
+
+- We have full flexibility in defining a record that matches our needs.
+- When people work with our internal Record it makes more sense because it is actually the same as an OpenCDC record.
+- If our record is marshalled into JSON (or avro or proto) it's OpenCDC compatible by default.
+
+#### Cons
+
+- We define a new format that nobody is using but us. https://xkcd.com/927/
+- Breaking change for connectors, the record will change.
+- Debezium compatibility has to be achieved through transforms.
+
+#### Changes to record
+
+We propose the following changes to the `Record` type:
+
+```diff
+type Record struct {
+	// Position uniquely represents the record.
+	Position Position
+	// Metadata contains additional information regarding the record.
+	Metadata map[string]string
++	// Operation defines if this record contains data related to an insert,
++	// update, delete or snapshot.
++	Operation Operation
+-	// CreatedAt represents the time when the change occurred in the source
+-	// system. If that's impossible to find out, then it should be the time the
+-	// change was detected by the connector.
+-	CreatedAt time.Time
+-	// Key represents a value that should be the same for records originating
+-	// from the same source entry (e.g. same DB row). In a destination Key will
+-	// never be null.
+-	Key Data
+-	// Payload holds the actual information that the record is transmitting. In
+-	// a destination Payload will never be null.
+-	Payload Data
++	// Before holds the key and payload data that was valid before the change.
++	// This field is only populated if the operation is update or delete.
++	Before struct {
++		// Key contains the data and schema of the key before the change.
++		Key DataWithSchema
++		// Payload contains the data and schema of the payload before the change.
++		Payload DataWithSchema
++	}
```

Comment:
I'm not myself sure about this, but I'll throw it to hear opinions: do we need metadata in the before and after structs? Metadata about the payload that is. For example, if we have an S3 source, I imagine the payload will be an S3 file itself. However, the S3 object also has metadata, which can change and in which we might be interested.

Reply:
Interesting thought, but I would argue that this kind of metadata is part of the payload. Given that we don't prescribe any specific structure for the payload itself the connector is welcome to create a metadata field in the payload itself, then changes to that metadata can be tracked if needed.

Reply:
Yup, that's true, keeping in the payload is definitely an option and has its pros for sure. A con though, is that it requires a "mini schema" in the connector to package the data and metadata togeher. It also adds to processing time for, for example, binary there, since it needs to be encoded.

Reply:
Before I say anything, this is the only thing (and it's minor) which I had for a comment. I like the rest of the OpenCDC, and AFAIC we can merge it. :muscle:  I don't think this is even close to being a blocker.

I too don't see many use cases either, which is a strong argument for not including it, and which is why I don't feel strongly about this. On the other hand, it's easier to remove an unused field, than to have client code work around the limitation of not having it.

What we would get with metadata in `before` and `after` is basically support for "context" changes, and not only changes in data objects themselves. 

> but any code dealing with the payload or key would need to somehow deal with the metadata as well

Hmm, why so? We may not be thinking about the same thing, but I'm thinking about a specific connector's code which is free to use or just ignore this field (in the same way it will ignore the `before` field in a number of cases). And when it comes to the code responsible for serialization/deserialization, then it would be just one more field.

> It would also create strange expectations 

Do we want to create expectations here about that all though? This document is detailed when it comes to how changes are described (kudos for that!), but we may not have to go into what triggers an update. That sounds like something source connectors should decide.

Reply:
I won't see this comment as a blocker then and merge the PR, but I don't want to leave you hanging so I will at least respond before doing so 🙂 

---

First, seems like we agree that such a field would have few use cases. Where we might have different opinions is if having such a field would create more problems or solve more problems. I'm thinking it would create more problems. Why? Complexity and expectations.

Complexity - here I mean mostly higher cognitive load for the developer. Having a rarely used field doesn't mean it can be ignored, it probably exists for a reason, some connectors will use it and will expect that metadata to have an effect. A developer will have to think about what to do with that field, decide if it's safe to ignore it, figure out what the difference is between record metadata and payload metadata. Unless the field would have a clear use case and a clearly defined purpose separating it from the record metadata, it could create more confusion.

Expectations - besides the expectations I mentioned in my previous comment I want to point out that having a field that can be safely ignored could create false expectations in the behavior across connectors. Say a connector puts some data in this metadata field while another connector ignores this field, this could mean they aren't as compatible as we would expect. If the metadata can be safely ignored it can be put into the record metadata. If it shouldn't be ignored it can be part of the payload. That are clear expectations. On the other hand, wondering if a field will be present in the record metadata or in the payload metadata and wondering how behavior will change in either case, that aren't clear expectations.

To sum up, I'd say the main reason for my opinion is the lack of a clear enough distinction between "record metadata", "payload metadata" and "payload data". The first one is metadata, that can safely be ignored, there is no expectation of the connector using these fields, though it can use it to alter its behavior (I would compare it to HTTP headers or gRPC metadata). The payload on the other hand is the "meat of the record", that's the content that caused for the record to be created in the first place. Adding "payload metadata" would put it somewhere in between and I just don't see how it would improve the user experience.